spec: updating ics27 spec to ensure parity with implementation (#616)
cosmos/ibc: spec/app/ics-027-interchain-accounts/README.md

### Motivation

> ICS-27 Interchain Accounts outlines a cross-chain account management protocol built upon IBC. ICS-27 enabled chains can programmatically create accounts on other ICS-27 enabled chains & control these accounts via IBC transactions (instead of signing with a private key). Interchain accounts retain all of the capabilities of a normal account (i.e. stake, send, vote) but instead are managed by a separate chain via IBC in a way such that the owner of the accounts retain full control over how the accounts behave. 

Reviewer: Can you clarify the last part of this sentence to indicate that there are two separate accounts involved (one on the controller chain, one on the controlled chain, or whatever the terminology is)?

> ICS-27 Interchain Accounts outlines a cross-chain account management protocol built upon IBC. ICS-27 enabled chains can programmatically create accounts on other ICS-27 enabled chains & control these accounts via IBC transactions (instead of signing with a private key). Interchain accounts retain all of the capabilities of a normal account (i.e. stake, send, vote) but instead are managed by a separate chain via IBC in a way such that the owner of the accounts retain full control over how the accounts behave. 

Author: > Can you clarify the last part of this sentence to indicate that there are two separate accounts involved (one on the controller chain, one on the controlled chain, or whatever the terminology is)?

I think it's important to be more clear on this, I agree. Perhaps, it is better to simply say that the controller chain is the owner of all accounts it registers on a host chain. In terms of the current implementation, the `owner-id` (which is in the portID on the controller chain) is arbitrary and does not need to be mapped to an account address per se but it is what we have been recommending. We have been dancing around the semantics of this for a while. In the end, it's always down to the controller chain to determine what the `owner-id` is exactly. 

The controller chain will always be responsible for building some level of authentication around the interchain accounts API (i.e. not exposing the API publically) and the 'owner' will often times not be an end-user but instead controlled via some on-chain protocol / governance. 

I think in a follow up PR we should change the semantics around this @AdityaSripal

- `Interchain Account Owner`: An account on the controller chain. Every interchain account on a host chain has a respective owner account on the controller chain. 
- `Controller Chain`: The chain registering and controlling an account on a host chain. The controller chain sends IBC packets to the host chain to control the account.
- `Host Chain`: The chain where the interchain account is registered. The host chain listens for IBC packets from a controller chain which should contain instructions (e.g. cosmos SDK messages) that the interchain account will execute.

The IBC handler interface & IBC relayer module interface are as defined in [ICS 25](../ics-025-handler-interface) and [ICS 26](../ics-026-routing-module), respectively.

### Desired Properties

- Permissionless 
- Fault tolerance: A controller chain must not be able to control accounts registered by other controller chains. For example, in the case of a fork attack on a controller chain, only the interchain accounts registered by the forked chain will be vulnerable.

> - The ordering of transactions sent to an interchain account on a host chain must be maintained. Transactions should be executed by an interchain account in the order in which they are sent by the controller chain.

Reviewer: What if transactions fail? Do we want a sort of causal execution dependency here, or is that out of scope?

> - If a channel closes, the controller chain must be able to regain access to registered interchain accounts by simply opening a new channel.

Reviewer: Really? Why, and how is this made safe?

Reviewer: I see that you tie accounts to the connection identifiers. I think this is fine, but I don't think it really gives any additional "fault liveness recovery", which I presume is what you want. If clients expire a new connection will need to be created in addition to new channels.

Reviewer 2: So we are considering two possibilities for the channel to close:

1. Timeout on ORDERED channel

In this case, the user can simply open a new channel on the same connection with the same port and that new channel will be able to control the original interchain account.

2. Underlying client expires or gets frozen for misbehaviour.

Here ICA does not have any in-protocol method for recovery, you are right. However, the underyling layers may have their own recovery mechanisms such that we don't have to create new connections/channels.
In fact, this is exactly the case in the ibc-go implementation. If a client expires or is frozen, then governance can come in and revive the client such that the connections/channels can resume working without having to create a whole new one.
So here, we can rely on governance to hopefully resolve the issue.
It's possible of course that governance does not choose to do this, in which case, there is nothing we can do.

Perhaps this reasoning should be written down succinctly @seantking

Reviewer: Right, OK, so the case that we're really concerned about here is (1), since that is the case where using connection identifiers allows for recovery in this case of timeouts of packets _but not_ the client. That makes sense to me. However, the semantics seem identical to (and more complicated than) an ordered channel which just doesn't close on timeouts, so maybe we should aim to add support for that instead and avoid the unnecessary extra handshake?

AFAIK (2) doesn't benefit from the possibility of using multiple channels so isn't relevant to this design question.

### Desired Properties

- Permissionless 
- Fault tolerance: A controller chain must not be able to control accounts registered by other controller chains. For example, in the case of a fork attack on a controller chain, only the interchain accounts registered by the forked chain will be vulnerable
- The ordering of transactions sent to an interchain account on a host chain must be maintained. Transactions should be executed by an interchain account in the order in which they are sent by the controller chain
- If a channel closes, the controller chain must be able to regain access to registered interchain accounts by simply opening a new channel

> - Each interchain account is owned by a single account on the controller chain. Only the owner account on the controller chain is authorized to control the interchain account. The controller chain is responsible for enforcing this logic

Reviewer 2: ```suggestion
- Each interchain account is owned by a single account on the controller chain. Only the owner account on the controller chain is authorized to control the interchain account. The controller chain is responsible for enforcing this logic.
```

// * CONTROLLER CHAIN *

// InitInterchainAccount is the entry point to registering an interchain account 

> // It generates a new port identifier using the owner address & connection identifiers

Reviewer 2: ```suggestion
// It generates a new port identifier using the owner address and connection identifiers
```

Due to how the mechanics of ICS-004 channel version negotiation operate the version passed into the host chain side (OnChanOpenTry) must match the counterparty version passed into the controller side (OnChanOpenAck) otherwise, there will be a resulting error at the IBC protocol level. 

Combined with the one channel per interchain account approach, this method of version negotiation allows us to pass the address of the interchain account back to the controller chain and create a mapping from controller port ID -> interchain account address during the `OnChanOpenAck` callback. As outlined in the [controlling flow](#Controlling-Flow), a controller chain will need to know the address of a registered interchain account in order to send transactions to the account on the host chain.

#### Version Negotiation Summary

| Initiator | Datagram         | Chain acted upon |  Version (Controller, Host) |
| --------- | ---------------- | ---------------- |  ---------------------- | 
| Controller| ChanOpenInit     | Controller       |  (ics27-1, none)        | 

> | Relayer   | ChanOpenTry      | Host             |  (ics27-1, ics27-1.{interchain-account-address})        | 

Reviewer 2: Define somewhere (either above or below this table), that `interchain-account-address` is the address on the host chain that is controlled by the owner through the given ICA channel

}

// Returns the ID of the active channel for a given portID, if present.
function GetActiveChannelID(portID: string) returns (string, boolean){
}

// Stores the address of the interchain account in state.
function SetInterchainAccountAddress(portID string, address string) returns (string) {
}

// Retrieves the interchain account from state.
function GetInterchainAccountAddress(portID string) returns (string, bool){
}

// DeleteActiveChannelID removes the active channel keyed by the provided portID stored in state
function (k Keeper) DeleteActiveChannelID(portID string) {
}

```

### Register & Controlling flows

#### Register Account Flow

To register an interchain account we require an off-chain process (relayer) to listen for `OnChannelOpenInit` events with the capability to finish a channel creation handshake on a given connection. 

1. The controller chain binds a new IBC port with an id composed of the *source/counterparty connection-ids* & the *interchain account owner address*

The IBC port identifier will look like this:
```
ics27-1.{connection-id}.{counterparty-connection-id}.{owner-account-address}
```

> This port will be used to create channels between the controller & host chain for a specific owner/interchain account pair. Only the account with `{owner-account-address}` matching the bound port will be authorized to send IBC packets over channels created with `ics27-1.{connection-id}.{counterparty-connection-id}.{owner-account-address}`. It is up to each controller chain to enforce this port registration and access on the controller side. 

Reviewer 3: ```suggestion
This port will be used to identify the channel(s) created between the controller & host chain for a specific owner/interchain account. Only the account with `{owner-account-address}` matching the bound port will be authorized to send IBC packets over channels created with port id `ics27-1.{connection-id}.{counterparty-connection-id}.{owner-account-address}`. It is up to each controller chain to enforce this port registration and access on the controller side. 
```

2. The controller chain emits an event signaling to open a new channel on this port given a connection 

> 3. A relayer listening for `OnChannelOpenInit` events will begin the channel creation handshake

Reviewer 3: ```suggestion
3. A relayer listening for `OnChannelOpenInit` events will continue the channel creation handshake
```

5. During the `OnChanOpenAck` callback on the controller chain a record of the interchain account address registered on the host chain during `OnChanOpenTry` is set in state with a mapping from portID -> interchain account address. See [version negotiation](#Version-Negotiation) section below for how to implement this
6. During the `OnChanOpenAck` & `OnChanOpenConfirm` callbacks on the controller & host chains respectively, the [active-channel](#Active-Channels) for this interchain account/owner pair, is set in state

#### Active Channels

The controller and host chain should keep track of an `active-channel` for each registered interchain account. The `active-channel` is set during the channel creation handshake process. This is a safety mechanism that allows a controller chain to regain access to an interchain account on a host chain in case of a channel closing. 

An example of an active channel on the controller chain can look like this:

```typescript
{
 // Controller Chain

>  SourcePortId: `ics27-<version>.<source-connection-id>.<counterparty-connection-id>.<owner-account-address>`,

Reviewer 3: It may be confusing to mix the "source" terminology with "counterparty". Consider using _**one**_ of these pairs of terms:

- `sourcePortId` and `destinationPortId`
- `portId` and `counterpartyPortId`

This is also in line with other ICS-es ([example](https://github.com/cosmos/ibc/tree/b25d9b8b2bae739821703404266256bb9a741027/spec/core/ics-004-channel-and-packet-semantics)).

 SourceChannelId: `<channel-id>`,
 // Host Chain
 CounterPartyPortId: `interchain-account`,
 CounterPartyChannelId: `<channel-id>`,
}
```

In the event of a channel closing, the active channel should be unset. ICS-27 channels should only be closed in the event of a timeout (if the implementation uses ordered channels) or in the unlikely event of a light client attack. Controller chains must retain the ability to open new ICS-27 channels and reset the active channel for a particular port id (containing owner-account-address) and associated interchain account. 

#### Version Negotiation

> ICS-27 takes advantage of [ISC-04 channel version negotiation](https://github.com/cosmos/ibc/tree/master/spec/core/ics-004-channel-and-packet-semantics#versioning) to store a mapping of the controller chain port ID to the newly registered interchain account address, on both the host chain & controller chains, during the channel creation handshake ([account registration flow](#Register-Account-Flow)). 

Reviewer 3: ```suggestion
ICS-27 takes advantage of [ICS-004 channel version negotiation](https://github.com/cosmos/ibc/tree/master/spec/core/ics-004-channel-and-packet-semantics#versioning) to store a mapping of the controller chain port ID to the newly registered interchain account address, on both the host chain & controller chains, during the channel creation handshake ([account registration flow](#Register-Account-Flow)). 
```

Due to how the mechanics of ICS-004 channel version negotiation operate the version passed into the host chain side (OnChanOpenTry) must match the counterparty version passed into the controller side (OnChanOpenAck) otherwise, there will be a resulting error at the IBC protocol level. 

Combined with the one channel per interchain account approach, this method of version negotiation allows us to pass the address of the interchain account back to the controller chain and create a mapping from controller port ID -> interchain account address during the `OnChanOpenAck` callback. As outlined in the [controlling flow](#Controlling-Flow), a controller chain will need to know the address of a registered interchain account in order to send transactions to the account on the host chain.

#### Version Negotiation Summary

> `interchain-account-address` is the address of the interchain account registered on the host chain by the controller chain.

Reviewer 3: Very clear!

| Initiator | Datagram         | Chain acted upon |  Version (Controller, Host) |
| --------- | ---------------- | ---------------- |  ---------------------- | 
| Controller| ChanOpenInit     | Controller       |  (ics27-1, none)        | 
| Relayer   | ChanOpenTry      | Host             |  (ics27-1, ics27-1.{interchain-account-address})        | 
| Relayer   | ChanOpenAck      | Controller       |  (ics27-1, ics27-1.{interchain-account-address})        | 
| Relayer   | ChanOpenConfirm  | Host             |  (ics27-1, ics27-1.{interchain-account-address})        | 

The channel `version` string passed into the `OnChannelOpenTry` callback by a relayer will contain the account address of the interchain account that will be registered on the host chain. A relayer will then pass this version into the `OnChannelOpenAcknowledgement` step on the controller chain as the `counterpartyVersion`. The controller chain will then proceed to parse the interchain account address from the `counterpartyVersion` string and store this address in state with a map to the associated port id.  

#### Controlling Flow

> Once an interchain account is registered on the host chain a controller chain can begin sending instructions (messages) to the host chain to control the account. 

Reviewer 3: What does _registered_ mean? Does it mean that `ChanOpenConfirm` executed at the host chain?

1. The controller chain calls `TrySendTx` and passes message(s) that will be executed on the host side by the associated interchain account (determined by the controller side port identifier)

> Cosmos SDK psuedo code example:

Reviewer 3: ```suggestion
Cosmos SDK pseudo-code example:
```

```typescript
interchainAccountAddress := GetInterchainAccountAddress(portId)
msg := &banktypes.MsgSend{FromAddress: interchainAccountAddress, ToAddress: ToAddress, Amount: amount}
icaPacketData = InterchainAccountPacketData{
   Type: types.EXECUTE_TX,
   Data: serialize(msg),
   Memo: "memo",
}

// Sends the message to the host chain, where it will eventually be executed 
TrySendTx(ownerAddress, connectionId, counterPartyConnectionId, data)
```

4. The host chain upon receiving the IBC packet will call `DeserializeTx`. 
    

> 5. The host chain will then call `ExecuteTx` & `AuthenticateTx` for each message and return an acknowledgment containing a success or error.  

Reviewer 3: ```suggestion
5. The host chain will then call `AuthenticateTx` and `ExecuteTx` for each message and return an acknowledgment containing a success or error.  
```

// * UTILITY FUNCTIONS *

> // Sets the active channel

Reviewer 2: for a given portID

function SetActiveChannelID(portId: string, channelId: string) returns (error){
}

> // Returns the id of the active channel if present

Reviewer 2: for a given portID

function GetActiveChannelID(portId: string) returns (string, boolean){
}

// Stores the address of the interchain account in state
function SetInterchainAccountAddress(portId string, address string) returns (string) {
}

// Gets the interchain account from state
function GetInterchainAccountAddress(portId string) returns (string, bool){
}
```

### Register & Controlling flows

#### Register Account Flow

To register an interchain account we require an off-chain process (relayer) to listen for `OnChannelOpenInit` events with the capability to finish a channel creation handshake on a given connection. 

1. The controller chain binds a new IBC port with an id composed of the *source/counterparty connection-ids* & the *interchain account owner address*

The IBC port identifier will look like this:
```
ics27-1.{connection-id}.{counterparty-connection-id}.{owner-address}
```

> This port will be used to create channels between the controller & host chain for a specific owner/interchain account pair. Only the account with `{owner-address}` matching the bound port will be authorized to send IBC packets over channels created with `ics27-1.{connection-id}.{counterparty-connection-id}.{owner-address}`. It is up to each controller chain to enforce this port registration and access on the controller side. 

Reviewer 2: `owner-address` is the account address on the controller chain, correct?

Author: Correct. I've been thinking lately that we should maybe use the terminology `owner-id` as it's up to the controller chain to manage this, and there's not any enforcement on the host side that this needs to be an account address. WDYT?

Reviewer 2: Sounds fair, so long as it is consistent

Author: I decided to keep this as the address of the account owner on the controller chain. I've replaced any reference of `owner-address` or `owner-id` with `owner-account-address`. 

I think the opening definition for `Interchain Account Owner` should make it obvious what the `owner-account-address` is representing. WDYT?

2. The controller chain emits an event signaling to open a new channel on this port given a connection 
3. A relayer listening for `OnChannelOpenInit` events will begin the channel creation handshake
4. During the `OnChanOpenTry` callback on the host chain an interchain account will be registered and a mapping of the interchain account address to the owner account address will be stored in state (this is used for authenticating transactions on the host chain at execution time). 
5. During the `OnChanOpenAck` callback on the controller chain a record of the interchain account address registered on the host chain during `OnChanOpenTry` is set in state with a mapping from portID -> interchain account address. See [version negotiation](#Version-Negotiation) section below for how to implement this

> 6. During the `OnChanOpenAck` & `OnChanOpenConfirm` callbacks on the controller & host chains respectively, the [active-channel](#Active-Channels) for this interchain account/owner pair, is set in state

Reviewer 2: Conclude each statement with `.`

#### Active Channels

The controller and host chain should keep track of an `active-channel` for each registered interchain account. The `active-channel` is set during the channel creation handshake process. This is a safety mechanism that allows a controller chain to regain access to an interchain account on a host chain in case of a channel closing. 

> An active channel on the controller chain will look like this:

Reviewer 2: ```suggestion
An example of an active channel on the controller chain can look like this:
```

Reviewer 2: This is not the struct for all active channels but just a single example

```typescript
{
 // Controller Chain

>  SourcePortId: `ics27-1.0.0.<owner-id>`,

Reviewer 2: Define the structure of `owner-id`

>  SourcePortId: `ics27-1.0.0.<owner-id>`,

Reviewer 2: You have `owner-id` in some places, and `owner-address` in some places. Standardize the terminology and define it to be the address on the controller chain that owns the interchain account.

Author: https://github.com/cosmos/ibc/pull/616#discussion_r747497839

Author: update: https://github.com/cosmos/ibc/pull/616#discussion_r750099314

>  SourcePortId: `ics27-1.0.0.<owner-id>`,

Reviewer 2: Also it doesn't make sense here to give an example `connection-id` and `counterparty-connection-id` but leave the owner address as a variable.

Either give them all an example value, or leave them all as variables

Author: Yeah, I agree, a good point. I removed the cosmos address as it was specific to the hub but I can put it back in.

Author: I actually put them all as variables in the end. :handshake:

Due to how the mechanics of ICS-004 channel version negotiation operate the version passed into the host chain side (OnChanOpenTry) must match the counterparty version passed into the controller side (OnChanOpenAck) otherwise there will be a resulting error at the IBC protocol level. 

Combined with the one channel per interchain account approach, this method of version negotiation allows us to pass the address of the interchain account back to the controller chain and create a mapping from controller port ID -> interchain account address during the `OnChanOpenAck` callback. As outlined in the [controlling flow](#Controlling-Flow), a controller chain will need to know the address of a registered interchain account in order to send transactions to the account on the host chain.

#### Version negotiation summary

| Initiator | Datagram         | Chain acted upon |  Version (Controller, Host) | Port ID (Controller, Host) |
| --------- | ---------------- | ---------------- |  ---------------------- | ---------------------- |
| Controller| ChanOpenInit     | Controller       |  (ics27-1, none)        | (ics27-1.0.0.<owner-id>, interchain-account)           |

> | Relayer   | ChanOpenTry      | Host             |  (ics27-1, ics27-1.<owner-id>)        | (ics27-1.0.0.<owner-id>, interchain-account)       |

Reviewer 2: This table seems incorrect. `owner-id` sounds like the account address on the controller chain since that is the "owner" of the interchain account.

But from my understanding, the version set by host chain contains the address of the host chain that is being controlled by the owner address on the controller chain.

I think the address in version string should have a different name: `host-address`?

Author: great catch! this is indeed incorrect. <owner-address> here should be the interchain account address registered on the host chain.

Author: I updated the variable name to be `interchain-account-address` as I'm hesitant to add further terminology here. Open to discussion. 

I also removed the port-id column as I noticed that I wasn't using variable names for the connection identifiers and adding them in makes the column difficult to parse. I think it's fine without.

Author: The reason I replaced the `<>` with `{}` is simply because using `<>` inside a table makes the content invisible on github README for some reason.

| Relayer   | ChanOpenAck      | Controller       |  (ics27-1, ics27-1.<owner-id>)        | (ics27-1.0.0.<owner-id>, interchain-account)        | (ics27-1.0.0.<owner-id>, interchain-account)        |
| Relayer   | ChanOpenConfirm  | Host             |  (ics27-1, ics27-1.<owner-id>)        | (ics27-1.0.0.<owner-id>, interchain-account)           | (ics27-1.0.0.<owner-id>, interchain-account)           |

> 

Reviewer 2: Briefly describe how controller chain can retrieve host address from counterparty version

Author: Done :+1:

#### Controlling Flow

Once an interchain account is registered on the host chain a controller chain can begin sending instructions (messages) to the host chain to control the account. 

1. The controller chain calls `TrySendTx` and passes message(s) that will be executed on the host side by the associated interchain account (determined by the controller side port identifier)

Cosmos SDK psuedo code example:

```typescript
interchainAccountAddress := GetInterchainAccountAddress(portId)
msg := &banktypes.MsgSend{FromAddress: interchainAccountAddress, ToAddress: ToAddress, Amount: amount}
icaPacketData = InterchainAccountPacketData{
   Type: types.EXECUTE_TX,
   Data: serialize(msg),
   Memo: "memo",
}

// Sends the message to the host chain, where it will eventually be executed 

> TrySendTx(chanCapability, ownerAddress, connectionId, counterPartyConnectionId, data)

Reviewer 2: ```suggestion
TrySendTx(ownerAddress, connectionId, counterPartyConnectionId, data)
```

The capabilities aren't mentioned in the spec since it is SDK specific authentication schema. Similar to how we don't include `ctx sdk.Context` in our spec documents.

I would just say that it is responsibility of ICA module to authenticate that the caller of `TrySendTx` is the owner of the interchain account

Author: Cool. Good to know :handshake:

function onChanOpenConfirm(
  portIdentifier: Identifier,
  channelIdentifier: Identifier) {
  // set the active channel for this owner/interchain account pair
  setActiveChannel(portIdentifier)
}
```

```typescript
function onChanCloseInit(
  portIdentifier: Identifier,
  channelIdentifier: Identifier) { \
    // unset the active channel for this owner/interchain account pair

>     DeleteActiveChannel(portIdentifier, channelIdentifier)

Reviewer 2: Shouldn't this return an error always? Ahh, is this because the ICA-auth may decide to close the channel? If so, then the line above about ICA channels only closing on timeouts should be changed

Author: This should return an error you are correct. I've updated this.

}
```

```typescript
function onChanCloseConfirm(
  portIdentifier: Identifier,
  channelIdentifier: Identifier) {
    // unset the active channel for this owner/interchain account pair
    DeleteActiveChannel(portIdentifier, channelIdentifier)
}
```

### Packet relay
`onRecvPacket` is called by the routing module when a packet addressed to this module has been received.

```typescript
function OnRecvPacket(packet Packet) {
    	ack := NewResultAcknowledgement([]byte{byte(1)})

	// only attempt the application logic if the packet data
	// was successfully decoded
	if ack.Success() {
		switch data.Type {
	          case types.EXECUTE_TX:
		        msgs, err := types.DeserializeTx(data.Data)
		        if err != nil {
			      return err
		        }

> 		        if err = executeTx(ctx, packet.SourcePort, packet.DestinationPort, packet.DestinationChannel, msgs); err != nil {

Reviewer 2: Does `executeTx` include authentication? If so that should be commented here.
